### PR TITLE
Use Arrow URI FileSystem backed instance to retrieve remote files

### PR DESCRIPTION
**Pull request by jdye64**
Arrow offers an API that allows for users to provide a uri definition for target files. This PR will use that api and create a new `arrow_io_source` constructor to accept that information from the user and then create the appropriate FileSystem instance and configure it for access to that file.

This closes: #7475

### CODE REVIEW

**Comment by jdye64** [comment]
Could someone please label this PR as [WIP]? I do not have the permissions to do so.

**Comment by kkraus14** [comment]
> Could someone please label this PR as [WIP]? I do not have the permissions to do so.

Can you convert the PR to a draft using a link on the right hand side?

**Comment by ayushdg** [comment]
In terms of configuration options for s3 and hdfs we currently support the following options: [hdfs](https://docs.dask.org/en/latest/remote-data-services.html#hadoop-file-system), [s3](https://docs.dask.org/en/latest/remote-data-services.html#amazon-s3).
Any ideas on how to expose the config options based on the filesystem class?

**Comment by vuule** [comment]
@jdye64 what is the status of this PR? Should it still target 21.06?

**Comment by jdye64** [comment]
@vuule lets move to 21.08. Having some build issues with it and a possible bug in Arrow that would need to be taken care of on the Arrow side first if true.

**Comment by jdye64** [comment]
rerun tests

**Comment by randerzander** [comment]
rerun tests

File: cpp/tests/io/arrow_io_source_test.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+auto const temp_env = static_cast<cudf::test::TempDirTestEnvironment*>(
+  ::testing::AddGlobalTestEnvironment(new cudf::test::TempDirTestEnvironment));
+
+// Base test fixture for tests
+struct ArrowIOTest : public cudf::test::BaseFixture {
+};
+
+TEST_F(ArrowIOTest, S3Filesystem)
+{
+  std::string s3_uri = "s3://ursa-labs-taxi-data/2010/06/data.parquet?region=us-east-2";
```
Comment (ayushdg) [inline cpp/tests/io/arrow_io_source_test.cpp:42]: Could this be updated to use a smaller file preferably hosted by rapids instead of a public dataset.

**Review by ayushdg — COMMENTED** [review]
Thanks for this @jdye64! Couple of comments regarding tests.

File: cpp/tests/io/arrow_io_source_test.cpp (hunk 1 above)
Comment (jdye64) [reply cpp/tests/io/arrow_io_source_test.cpp:69]: That would be great! Do we have an existing parquet file in our rapids public dataset? If so could you point me to that?

File: cpp/cmake/thirdparty/CUDF_GetArrow.cmake [hunk 2]
```diff
@@ -123,4 +124,4 @@ endfunction()
 
 set(CUDF_VERSION_Arrow 1.0.1)
 
-find_and_configure_arrow(${CUDF_VERSION_Arrow} ${CUDF_USE_ARROW_STATIC})
+find_and_configure_arrow(${CUDF_VERSION_Arrow} ${CUDF_USE_ARROW_STATIC})
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/CUDF_GetArrow.cmake:127]: Missing a newline at the end of the file.

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: cpp/tests/io/arrow_io_source_test.cpp (hunk 1 above)
Comment (jdye64) [reply cpp/tests/io/arrow_io_source_test.cpp:68]: https://github.com/rapidsai/ops/issues/1630

File: cpp/cmake/thirdparty/CUDF_GetArrow.cmake (hunk 2 above)
Comment (jdye64) [reply cpp/cmake/thirdparty/CUDF_GetArrow.cmake]: updated in latest commit

File: cpp/include/cudf/io/datasource.hpp [hunk 3]
```diff
@@ -302,6 +306,31 @@ class arrow_io_source : public datasource {
   };
 
  public:
+  /**
+   * @brief Constructs an object from an Apache Arrow Filesystem URI
+   *
+   * @param Apache Arrow Filesystem URI
+   */
+  explicit arrow_io_source(std::string arrow_uri)
```
Comment (jrhemstad) [inline cpp/include/cudf/io/datasource.hpp:314]:
```suggestion
  explicit arrow_io_source(std::string_view arrow_uri)
```

File: cpp/include/cudf/io/datasource.hpp [hunk 4]
```diff
@@ -302,6 +306,31 @@ class arrow_io_source : public datasource {
...
+      arrow::fs::FileSystemFromUri(arrow_uri);
+    CUDF_EXPECTS(result.ok(), "Failed to generate Arrow Filesystem instance from URI.");
+    filesystem = result.ValueOrDie();
+
+    // Parse the path from the URI
+    size_t start = arrow_uri.find(uri_start_delimiter) == std::string::npos
+                     ? 0
+                     : arrow_uri.find(uri_start_delimiter) + uri_start_delimiter.size();
+    size_t end       = arrow_uri.find(uri_end_delimiter) - start;
+    std::string path = arrow_uri.substr(start, end);
```
Comment (jrhemstad) [inline cpp/include/cudf/io/datasource.hpp:326]:
```suggestion
    std::string_view path = arrow_uri.substr(start, end);
```

File: cpp/include/cudf/io/datasource.hpp [hunk 5]
```diff
@@ -340,6 +369,9 @@ class arrow_io_source : public datasource {
   }
 
  private:
+  const std::string uri_start_delimiter = "//";
+  const std::string uri_end_delimiter   = "?";
+  std::shared_ptr<arrow::fs::FileSystem> filesystem;
```
Comment (jrhemstad) [inline cpp/include/cudf/io/datasource.hpp:374]: Why does `filesystem` need to be a data member? It looks like it isn't used anywhere other than in the ctor.

File: cpp/include/cudf/io/datasource.hpp [hunk 6]
```diff
@@ -340,6 +369,9 @@ class arrow_io_source : public datasource {
   }
 
  private:
+  const std::string uri_start_delimiter = "//";
+  const std::string uri_end_delimiter   = "?";
```
Comment (jrhemstad) [inline cpp/include/cudf/io/datasource.hpp:373]: If these are only used in the ctor, I think it'd be better to make these local variables in the ctor.

File: cpp/include/cudf/io/datasource.hpp (hunk 5 above)
Comment (jdye64) [reply cpp/include/cudf/io/datasource.hpp:374]: Otherwise it goes out of scope and subsequent `read_host` calls result in segmentation faults

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Looks good overall.

File: cpp/include/cudf/io/datasource.hpp (hunk 5 above)
Comment (jrhemstad) [reply cpp/include/cudf/io/datasource.hpp:374]: Ah, I see.

**Review by jrhemstad — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Comment by jdye64** [comment]
rerun tests

**Comment by jdye64** [comment]
rerun tests

**Pull request merged**